### PR TITLE
Add test for TeamPolicy array reduction and fix Cuda and HIP

### PR DESCRIPTION
**Pull request by masterleinad**
In reference to https://github.com/kokkos/kokkos/pull/6293#issue-1810833445. It seems that we don't test array reductions for TeamPolicy in the CI. This pull request copies a very similar test to use array reductions instead of a C-style array of fixed size.
It also fixes the respective implementation for Cuda and HIP.

### CODE REVIEW

**Comment by fnrizzi** [comment]
why is this PR also touching impl for CUDA and HIP too?

File: core/unit_test/TestTeam.hpp [hunk 1]
```diff
@@ -290,6 +290,58 @@ class ReduceTeamFunctor {
   }
 };
 
+template <typename ScalarType, class DeviceType, class ScheduleType>
```
Comment (ldh4) [inline core/unit_test/TestTeam.hpp:293]: Just nitpicking. Can these be either all `typename`s or `class`es?

**Review by ldh4 — APPROVED** [review]

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp [hunk 2]
```diff
@@ -711,16 +711,17 @@ class ParallelReduce<CombinedFunctorReducerType,
...
                                     ReducerType::static_value_size() /
                                         sizeof(word_size_type)>
         word_count(m_functor_reducer.get_reducer().value_size() /
                    sizeof(word_size_type));
 
-    reference_type value = m_functor_reducer.get_reducer().init(
-        kokkos_impl_cuda_shared_memory<word_size_type>() +
-        threadIdx.y * word_count.value);
+    reference_type value =
+        m_functor_reducer.get_reducer().init(reinterpret_cast<pointer_type>(
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp:722]: So it is not so much a "adding test", the code was broken wasn't it?

File: core/unit_test/TestTeam.hpp [hunk 3]
```diff
@@ -290,6 +290,58 @@ class ReduceTeamFunctor {
...
+
+  KOKKOS_INLINE_FUNCTION
+  void join(value_type dst, const value_type src) const {
+    dst[0] += src[0];
+    dst[1] += src[1];
+    dst[2] += src[2];
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const typename policy_type::member_type ind,
```
Comment (dalg24) [inline core/unit_test/TestTeam.hpp:327]: Not blocking but I am really not a fan of calling it "ind".
Also why are we not passing by ref?

File: core/unit_test/TestTeam.hpp [hunk 4]
```diff
@@ -304,39 +356,73 @@ class TestReduceTeam {
...
+        result_type tmp(&result[i][0]);
+        Kokkos::parallel_reduce(team_exec, functor_type(nwork), tmp);
+      }
+
+      execution_space().fence();
+
+      for (unsigned i = 0; i < Repeat; ++i) {
+        for (unsigned j = 0; j < Count; ++j) {
+          const uint64_t correct = 0 == j % 3 ? nw : nsum;
+          ASSERT_EQ((ScalarType)correct, result[i][j]);
```
Comment (dalg24) [inline core/unit_test/TestTeam.hpp:424]: This feels silly, just inline the assertion and inject more information with a `<< <helpful message>;`

File: core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp:722]: Well, yeah, it was broken for `Cuda` and `HIP` but not for the other backends AFAIK which means we didn't have a test.

File: core/unit_test/TestTeam.hpp [hunk 5]
```diff
@@ -290,6 +290,58 @@ class ReduceTeamFunctor {
...
+class ArrayReduceTeamFunctor {
+ public:
+  using execution_space = DeviceType;
+  using policy_type     = Kokkos::TeamPolicy<ScheduleType, execution_space>;
+  using size_type       = typename execution_space::size_type;
+
+  using value_type      = ScalarType[];
+  size_type value_count = 3;
+
+  const size_type nwork;
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/TestTeam.hpp:303]: All this is copied from `ReduceTeamFunctor` and `TeamFunctor` (array reduction for `RangePolicy`).

File: core/unit_test/TestTeam.hpp [hunk 6]
```diff
@@ -304,39 +348,73 @@ class TestReduceTeam {
...
+      for (unsigned i = 0; i < Repeat; ++i) {
+        result_type tmp(&result[i][0]);
+        Kokkos::parallel_reduce(team_exec, functor_type(nwork), tmp);
+      }
+
+      execution_space().fence();
+
+      for (unsigned i = 0; i < Repeat; ++i) {
+        for (unsigned j = 0; j < Count; ++j) {
+          ASSERT_EQ(j % 3 ? nsum : nw, static_cast<uint64_t>(result[i][j]))
```
Comment (Rombur) [inline core/unit_test/TestTeam.hpp:415]: Why do you have a modulo 3? The test will blow up if `j` is greater or equal to 3.

Comment (masterleinad) [reply core/unit_test/TestTeam.hpp]: I replaced it.

File: core/unit_test/TestTeam.hpp [hunk 7]
```diff
@@ -304,39 +348,73 @@ class TestReduceTeam {
...
+      for (unsigned i = 0; i < Repeat; ++i) {
+        result_type tmp(&result[i][0]);
+        Kokkos::parallel_reduce(team_exec, functor_type(nwork), tmp);
+      }
+
+      execution_space().fence();
+
+      for (unsigned i = 0; i < Repeat; ++i) {
+        for (unsigned j = 0; j < Count; ++j) {
+          ASSERT_EQ(j == 0 ? nsum : nw, static_cast<uint64_t>(result[i][j]))
```
Comment (Rombur) [inline core/unit_test/TestTeam.hpp:415]: This is wrong. The old code was 
```
ASSERT_EQ(j % 3 ? nsum : nw, static_cast<uint64_t>(result[i][j]))
```
You can just use 
```
ASSERT_EQ(j ? nsum : nw, static_cast<uint64_t>(result[i][j]))
```

**Review by Rombur — APPROVED** [review]

File: core/unit_test/TestTeam.hpp [hunk 8]
```diff
@@ -304,41 +348,78 @@ class TestReduceTeam {
...
+      for (unsigned i = 0; i < Repeat; ++i) {
+        for (unsigned j = 0; j < Count; ++j) {
+          const uint64_t correct = (j == 0) ? nw : nsum;
+          ASSERT_EQ((ScalarType)correct, result[i].value[j]);
+        }
+      }
     }
 
-    execution_space().fence();
+// FIXME_MSVC FIXME_32BIT Test is known to fail
```
Comment (dalg24) [inline core/unit_test/TestTeam.hpp:390]: Fail or not compile?
If that is fail why are we not skipping instead of guarding the whole thing?

Comment (masterleinad) [reply core/unit_test/TestTeam.hpp:390]: Mostly, because the output would be misleading since only have the test is skipped.

Comment (dalg24) [reply core/unit_test/TestTeam.hpp:390]: Can/should these be split into two unit tests?  (Haven't looked into details)

File: core/unit_test/TestTeamReductionScan.hpp [hunk 9]
```diff
@@ -52,18 +56,70 @@ TEST(TEST_CATEGORY, team_double_reduce) {
...
+TEST(TEST_CATEGORY, team_long_array_reduce) {
+// FIXME_MSVC FIXME_32BIT Test is known to fail
+#if !defined(KOKKOS_COMPILER_MSVC) && !defined(KOKKOS_IMPL_32BIT)
+  GTEST_SKIP() << "Test know to fail for MSVC or 32-bit builds";
+#endif
+
+#ifdef KOKKOS_ENABLE_OPENMPTARGET  // FIXME_OPENMPTARGET: Not implemented
+  if constexpr (!std::is_same<TEST_EXECSPACE,
+                              Kokkos::Experimental::OpenMPTarget>::value)
+#endif
```
Comment (dalg24) [inline core/unit_test/TestTeamReductionScan.hpp:83]: Uggg
The gtest output will suggest the test run.  Not blocking considering the issue was already there

File: core/unit_test/TestTeamReductionScan.hpp [hunk 10]
```diff
@@ -52,18 +56,70 @@ TEST(TEST_CATEGORY, team_double_reduce) {
...
+    TestReduceTeam<long, TEST_EXECSPACE, Kokkos::Schedule<Kokkos::Static> >{}
+        .run_array_test(100000);
+    TestReduceTeam<long, TEST_EXECSPACE, Kokkos::Schedule<Kokkos::Dynamic> >{}
+        .run_array_test(100000);
+  }
+}
+
+TEST(TEST_CATEGORY, team_double_array_reduce) {
+// FIXME_MSVC FIXME_32BIT Test is known to fail
+#if defined(KOKKOS_COMPILER_MSVC) && !defined(KOKKOS_IMPL_32BIT)
```
Comment (dalg24) [inline core/unit_test/TestTeamReductionScan.hpp:102]:
```suggestion
#if !defined(KOKKOS_COMPILER_MSVC) && !defined(KOKKOS_IMPL_32BIT)
```

Comment (masterleinad) [reply core/unit_test/TestTeamReductionScan.hpp]: Thanks!

**Comment by dalg24** [comment]
Same for the MSVC build
```
[ RUN      ] serial.team_long_array_reduce
C:\projects\source\core\unit_test\TestTeam.hpp(421): error: Expected equality of these values:
  j ? nsum : nw
    Which is: 5000050000
  static_cast<uint64_t>(result[i][j])
    Which is: 705082704
failing at repeat 0 and index 1
C:\projects\source\core\unit_test\TestTeam.hpp(421): error: Expected equality of these values:
  j ? nsum : nw
    Which is: 5000050000
  static_cast<uint64_t>(result[i][j])
    Which is: 705082704
failing at repeat 0 and index 1
[  FAILED  ] serial.team_long_array_reduce (1076 ms)

**Comment by dalg24** [comment]
OK the macro guard was opposite of what we wanted.  I see you fixed it now.

**Pull request merged**